spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/AbstractFileRewriteCoordinator.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.ContentFile;
+import org.apache.iceberg.HasTableOperations;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractFileRewriteCoordinator<F extends ContentFile<F>> {
```

Comment:
A lot has changed since we added `FileScanTaskSetManager`. I think that was even me? Back then we did not have any other scan task types and our Spark logic wasn't that flexible in terms of supporting those scan task types. That's no longer true.

Instead of offering a file rewrite coordinator for each file type and adapting the existing logic, I'd deprecate existing classes and create a new more flexible path. We will eventually drop the old path.

I'd start by adding a more generic read option.

```
public static final String SCAN_TASK_SET_ID = "scan-task-set-id";
```

We would then create a new task set manager with signature like below (just an idea, can be different).

```
public class ScanTaskSetManager {

  public void stageTasks(Table table, String setId, List<? extends ScanTask> tasks) {
    ...
  }

  @SuppressWarnings("unchecked")
  public <T extends ScanTask> List<T> fetchTasks(Table table, String setId) {
    ...
  }

  @SuppressWarnings("unchecked")
  public <T extends ScanTask> List<T> removeTasks(Table table, String setID) {
    ...
  }
}
```

Then `SparkStagedScanBuilder` and `SparkStagedScan` classes that would support arbitrary scan tasks. The scan implementation will be very straightforward.

```
@Override
protected List<ScanTaskGroup<ScanTask>> taskGroups() {
  if (taskGroups == null) {
    ScanTaskSetManager taskSetManager = ScanTaskSetManager.get();
    List<ScanTask> tasks = taskSetManager.fetchTasks(table(), taskSetId);
    ValidationException.check(
        tasks != null,
        "Task set manager has no tasks for table %s with task set ID %s",
        table().name(),
        taskSetId);

    this.taskGroups = TableScanUtil.planTaskGroups(tasks, splitSize, splitLookback, openFileCost);
  }

  return taskGroups;
}
```

Then new path can be used for data and delete file compaction.

Reply:
This may seem like more work but it will actually make this PR smaller and it is very close to what you did initially.

Reply:
We can keep existing `SparkFilesScanBuilder` and `SparkFilesScan` untouched for now. Just deprecate.

Reply:
Done, let me know if its what you mean.

Deprecated SparkFilesScanBuilder and SparkFilesScan.  Can we remove them, and the support for the existing flag "file-scan-task-set-id"?

Reply:
(note: I think the comment was for another class only?, so I did not change this class file)

Reply:
My bad. I commented on the wrong class. The rewrite coordinator belongs to writes. Can we include it in the follow-up change for writes? It will be easier to review it there. Let's focus on reading staged delete files in this PR.

Reply:
Removed this class